File: python/order_book.py
import heapq

class OrderBook:
    def __init__(self):
        self.orders = {}                   # order id: (price, quantity)
        self.price_to_quantity = {}        # price: quantity of orders at given price
        self.heap = []                     # max heap of prices

    def add_order(self, order_id, price, quantity):
        # O(log n)
        if order_id not in self.orders:
            self.orders[order_id] = (price, quantity)
            if price not in self.price_to_quantity:
                heapq.heappush_max(self.heap, price)
                self.price_to_quantity[price] = quantity
            else:
                self.price_to_quantity[price] += quantity

    def cancel_order(self, order_id):
        # O(1)
        if order_id not in self.orders:
            return False
        price, quantity = self.orders[order_id]
        self.price_to_quantity[price] -= quantity
        if self.price_to_quantity[price] == 0:
            del self.price_to_quantity[price]
        del self.orders[order_id]
        return True

    def get_best_price(self):
        # O(1)
        if not self.orders:
            return None
        while self.heap:
            if self.heap[0] in self.price_to_quantity:
                return self.heap[0]
            else:
                heapq.heappop_max(self.heap)
        return None

    def get_total_quantity_at_price(self, price):
        # O(1)
        if price not in self.price_to_quantity:
            return 0
        return self.price_to_quantity[price]
    
    def update_order(self, order_id, new_price=None, new_quantity=None):
        # O(1)
        if order_id not in self.orders:
            return False
        old_price, old_quantity = self.orders[order_id]
        if new_price is not None:
            if new_price not in self.price_to_quantity:
                heapq.heappush_max(self.heap, new_price)
                self.price_to_quantity[new_price] = 0
            if new_quantity is not None:
                self.orders[order_id] = (new_price, new_quantity)
                self.price_to_quantity[new_price] += new_quantity
            else:
                self.orders[order_id] = (new_price, old_quantity)
                self.price_to_quantity[new_price] += old_quantity
            self.price_to_quantity[old_price] -= old_quantity
        elif new_quantity is not None:
            self.orders[order_id] = (old_price, new_quantity)
            self.price_to_quantity[old_price] -= old_quantity
            self.price_to_quantity[old_price] += new_quantity
        if self.price_to_quantity[old_price] == 0:
            del self.price_to_quantity[old_price]
        return True
    
    def get_top_k_prices(self, k):
        # O(k log n)
        result = []
        temp = []
        while self.heap and len(result) < k:
            top_price = heapq.heappop_max(self.heap)
            if top_price in self.price_to_quantity:
                result.append(top_price)
                temp.append(top_price)
        for price in temp:
            heapq.heappush_max(self.heap, price)
        return result
        
    
order = OrderBook()
order.add_order(1, 50, 300)
order.add_order(2, 55, 250)
order.add_order(3, 60, 500)
order.add_order(4, 50, 1000)
print(order.get_best_price())
order.cancel_order(3)
print(order.get_best_price())
print(order.get_total_quantity_at_price(50))
print(order.update_order(4, 70, 1500))
print(order.get_best_price())
print(order.get_total_quantity_at_price(50))
order.add_order(5, 55, 100)
order.add_order(6, 55, 60)
print(order.get_top_k_prices(3))

File: cpp/order_book.cpp
#include <iostream>
#include <map>
#include <stdexcept>

enum Side { BUY, SELL };

struct Order {
	inline static std::atomic<size_t> next_id = 1;
	size_t id;
	double price;
	size_t quantity;
	Side side;
	Order* next;
	Order* prev;

public:
	Order(double price, size_t quantity, Side side, Order* next, Order* prev) {
		this->id = next_id++;
		this->price = price;
		this->quantity = quantity;
		this->side = side;
		this->next = next;
		this->prev = prev;
	}
};

struct PriceLevel {
	double price;
	size_t volume;
	Order* head;
	Order* tail;

public:
	PriceLevel(double price, size_t volume, Order* head, Order* tail) {
		this->price = price;
		this->volume = volume;
		this->head = head;
		this->tail = tail;
	}
};

typedef std::map<double, PriceLevel> prices_map;
class OrderBook {
	prices_map buy_orders;
	prices_map sell_orders;
	std::unordered_map<size_t, std::unique_ptr<Order>> orders;
	void remove_from_price_level(Order* order, PriceLevel& price_level, prices_map& map) {
		if (order->next != nullptr) {
			Order* next_order = order->next;
			if (order->prev != nullptr) {
				Order* prev_order = order->prev;
				prev_order->next = next_order;
				next_order->prev = prev_order;
			} else {
				next_order->prev = nullptr;
				price_level.head = next_order;
			}
		} else if (order->prev != nullptr) {
			Order* prev_order = order->prev;
			prev_order->next = nullptr;
			price_level.tail = prev_order;
		}
		price_level.volume -= order->quantity;
		if (price_level.volume == 0) {
			map.erase(order->price);
		}
	}

public:
	void place_order(double price, size_t quantity, Side side) {
		std::unique_ptr<Order> order_ptr = std::make_unique<Order>(price, quantity, side, nullptr, nullptr);
		Order* order = order_ptr.get();
		orders[order->id] = std::move(order_ptr);
		prices_map& map = (side == BUY) ? buy_orders : sell_orders;
		auto it = map.find(price);
		if (it == map.end()) {
			// new price level
			map.emplace(price, PriceLevel(price, quantity, order, order));
		} else {
			// existing price level
			PriceLevel& price_level = it->second;
			price_level.volume += quantity;
			Order* tail = price_level.tail;
			tail->next = order;
			order->prev = tail;
			price_level.tail = order;
		}
		std::cout << "Your order is Order # : " << order->id << std::endl;
	}
	double get_top(Side side) {
		if (side == BUY) {
			return buy_orders.rbegin()->first;
		} else {
			return sell_orders.begin()->first;
		}
	}
	void cancel_order(size_t id) {
		auto it = orders.find(id);
		Order* order = (it->second).get();
		prices_map& map = (order->side == BUY) ? buy_orders : sell_orders;
		PriceLevel& price_level = map.find(order->price)->second;
		remove_from_price_level(order, price_level, map);
		order->next = nullptr;
		order->prev = nullptr;
		orders.erase(it);
	}
	void update_order(size_t id, double new_price, size_t new_quantity) {
		Order* order = (orders.find(id)->second).get();
		if (order->price == new_price && order->quantity == new_quantity)
			return;
		prices_map& map = (order->side == BUY) ? buy_orders : sell_orders;
		PriceLevel& old_price_level = map.find(order->price)->second;
		remove_from_price_level(order, old_price_level, map);
		order->price = new_price;
		order->quantity = new_quantity;
		auto it = map.find(new_price);
		if (it == map.end()) {
			order->prev = nullptr;
			order->next = nullptr;
			map.emplace(new_price, PriceLevel(new_price, new_quantity, order, order));
		} else {
			PriceLevel& new_price_level = it->second;
			new_price_level.volume += new_quantity;
			Order* tail = new_price_level.tail;
			tail->next = order;
			order->prev = tail;
			order->next = nullptr;
			new_price_level.tail = order;
		}
	}
	size_t get_quantity_at_price(double price) {
		prices_map& map = (price >= sell_orders.begin()->first) ? sell_orders : buy_orders;
		auto it = map.find(price);
		if (it == map.end()) {
			throw std::invalid_argument(std::format("No orders at price {}", price));
		}
		PriceLevel& price_level = it->second;
		return price_level.volume;
	}
	std::vector<double> get_top_k_prices(int k, Side side) {
		prices_map& map = (side == BUY) ? buy_orders : sell_orders;
		std::vector<double> top_k_prices;
		if (side == BUY) {
			auto end = std::next(map.rbegin(), k);
			for (auto it = map.rbegin(); it != end; ++it) {
				top_k_prices.push_back(it->first);
			}
		} else if (side == SELL) {
			auto end = std::next(map.begin(), k);
			for (auto it = map.begin(); it != end; ++it) {
				top_k_prices.push_back(it->first);
			}
		}
		return top_k_prices;
	}
};

int main() {
	OrderBook order_book;
	order_book.place_order(33.34, 20, BUY);
	order_book.place_order(33.34, 30, BUY);
	order_book.place_order(33.36, 50, BUY);
	order_book.place_order(33.40, 20, SELL);
	order_book.place_order(33.39, 20, SELL);
	std::cout << order_book.get_top(BUY) << std::endl;
	std::cout << order_book.get_top(SELL) << std::endl;
	order_book.cancel_order(3);
	std::cout << order_book.get_top(BUY) << std::endl;
	order_book.update_order(2, 33.37, 100);
	std::cout << order_book.get_top(BUY) << std::endl;
	order_book.place_order(33.34, 50, BUY);
	std::cout << order_book.get_quantity_at_price(33.34) << std::endl;
	std::vector<double> top_k_prices = order_book.get_top_k_prices(2, BUY);
	std::string s;
	for (double price : top_k_prices) {
		s += std::to_string(price) + " ";
	}
	std::cout << s << std::endl;
	std::vector<double> top_k_prices_2 = order_book.get_top_k_prices(2, SELL);
	std::string k;
	for (double price : top_k_prices_2) {
		k += std::to_string(price) + " ";
	}
	std::cout << k << std::endl;
	std::cout << order_book.get_quantity_at_price(33.39) << std::endl;
	return 0;
}
